### PR TITLE
Fix the batch pragma suppression fixer to handle the case where multi…

### LINKED ISSUE #6455: Suppressing multiple compiler warnings on same token doesn't suppress them all
- Create a C# class library with following code:
```c#
namespace ClassLibrary1
{
    public class Class1
    {
        public static bool operator ==(Class1 c1, Class1 c2)
        {
            return true;
        }

        public static bool operator !=(Class1 c1, Class1 c2)
        {
            return true;
        }
    }
}
```

- Open error list and select both the warnings and attempt to suppress both of them in source:
```
Severity	Code	Description	Project	File	Line	Suppression State
Warning	CS0660	'Class1' defines operator == or operator != but does not override Object.Equals(object o)	ClassLibrary1	c:\users\mavasani\documents\visual studio 2015\Projects\ClassLibrary1\ClassLibrary1\Class1.cs	3	Active
Warning	CS0661	'Class1' defines operator == or operator != but does not override Object.GetHashCode()	ClassLibrary1	c:\users\mavasani\documents\visual studio 2015\Projects\ClassLibrary1\ClassLibrary1\Class1.cs	3	Active
```

- 
  **Expected:** Both diagnostics get suppressed
  **Actual:** Only one diagnostic (CS0660) gets suppressed

- **Expected fix:**
```c#
namespace ClassLibrary1
{
#pragma warning disable CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)
#pragma warning disable CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
    public class Class1
#pragma warning restore CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
#pragma warning restore CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)
    {
        public static bool operator ==(Class1 c1, Class1 c2)
        {
            return true;
        }

        public static bool operator !=(Class1 c1, Class1 c2)
        {
            return true;
        }
    }
}
```
  **Actual fix:**
```c#
namespace ClassLibrary1
{
#pragma warning disable CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)
    public class Class1
#pragma warning disable CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
#pragma warning restore CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)
    {
#pragma warning restore CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
        public static bool operator ==(Class1 c1, Class1 c2)
        {
            return true;
        }

        public static bool operator !=(Class1 c1, Class1 c2)
        {
            return true;
        }
    }
}
```

### PR DESCRIPTION
**Pull request by mavasani**
…ple diagnostics with different IDs are being suppressed on the same token

Current logic in batch pragma fixer merges individual fixes in a document by applying each pragma fix sequentially and keeping track of new diagnostics spans for remaining diagnostics. The latter logic that computes new spans was using out of date spans, causing the subsequent fixes to be incorrectly applied. This change fixes that logic.

Fixes #6455

### CODE REVIEW

**Comment by mavasani** [comment]
@heejaechang @srivatsn @basoundr

**Comment by mavasani** [comment]
@dotnet/roslyn-infrastructure prtest/win/dbg/unit64 seems to have timed out during build. Does someone want to jump onto the machine and see if the hung process is still around, and if so get a dump?

File: src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaBatchFixHelpers.cs [hunk 1]
```diff
@@ -123,71 +126,59 @@ private static class PragmaBatchFixHelpers
...
-                }
-            }
-
-            private static void UpdateDiagnosticSpans(ImmutableArray<Diagnostic> diagnostics, Dictionary<Diagnostic, TextSpan> currentDiagnosticSpans, TextChange textChange)
-            {
-                var isAdd = textChange.Span.Length == 0;
-                Func<TextSpan, bool> isPriorSpan = span => span.End <= textChange.Span.Start;
-                Func<TextSpan, bool> isFollowingSpan = span => span.Start >= textChange.Span.End;
-                Func<TextSpan, bool> isEnclosingSpan = span => span.Contains(textChange.Span);
+                Func<TextSpan, TextChange, bool> isPriorSpan = (span, textChange) => span.End <= textChange.Span.Start;
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaBatchFixHelpers.cs]: TextSpan, TextSpan? it doesnt actually use any information except span from TextChange.

File: src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaBatchFixHelpers.cs [hunk 2]
```diff
@@ -123,71 +126,59 @@ private static class PragmaBatchFixHelpers
...
+                    TextSpan originalSpan;
+                    if (!currentDiagnosticSpans.TryGetValue(diagnostic, out originalSpan))
                     {
-                        // Prior span, needs no update.
                         continue;
                     }
 
-                    var delta = textChange.NewText.Length - textChange.Span.Length;
-                    if (delta != 0)
+                    var currentSpan = originalSpan;
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaBatchFixHelpers.cs]: this seems useful library code.

File: src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs [hunk 3]
```diff
@@ -126,6 +124,24 @@ internal static SyntaxToken GetNewStartTokenWithAddedPragma(SyntaxToken startTok
                 return startToken.WithLeadingTrivia(trivia.InsertRange(index, pragmaTrivia));
             }
 
+            private static bool IsEndOfLineOrHasLeadingEndOfLine(SyntaxTrivia trivia, AbstractSuppressionCodeFixProvider fixer)
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs]: why isEndOfLine logic is in fixer?

File: src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs [hunk 4]
```diff
@@ -126,6 +124,24 @@ internal static SyntaxToken GetNewStartTokenWithAddedPragma(SyntaxToken startTok
...
+            {
+                return fixer.IsEndOfLine(trivia) ||
+                    (trivia.HasStructure && fixer.IsEndOfLine(trivia.GetStructure().DescendantTrivia().FirstOrDefault()));
+            }
+
+            private static bool IsEndOfLineOrHasTrailingEndOfLine(SyntaxTrivia trivia, AbstractSuppressionCodeFixProvider fixer)
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs]: these seems utility code as well.

**Comment by heejaechang** [comment]
:+1:

File: src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs (hunk 3 above)
Comment (mavasani) [inline src/Features/Core/Portable/CodeFixes/Suppression/AbstractSuppressionCodeFixProvider.PragmaHelpers.cs]: We should probably just move it out of language-specific fixer and use syntax facts service - will do it as part of next fix :)

**Comment by mavasani** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please

**Comment by mavasani** [comment]
I am going to fix another suppression related [bug](https://github.com/dotnet/roslyn/issues/6228) soon, will address all refactoring suggestions with that fix.

**Pull request merged**